Add callbacks to ServerSideRenderer to handle failures with custom renderers

mikejolley opened the pull request:
## Description
Added 3 optional callbacks to the `ServerSideRender` component so that consumers can change what is displayed based on the result. These include:

- ~~`onEmptyResponse`~~ -> emptyResponsePlaceholder
- ~~`onErrorResponse`~~ -> errorResponsePlaceholder
- ~~`onNullResponse`~~ -> loadingResponsePlaceholder

## How has this been tested?
Testing this with my own code/callback function.

```jsx
<ServerSideRender block={ name } attributes={ attributes } emptyResponsePlaceholder={ this.emptyResponsePlaceholder } />
```

And my callback:

```
	emptyResponsePlaceholder() {
		return (
			<Placeholder
				icon="category"
				label={ __( 'Products by Category', 'woo-gutenberg-products-block' ) }
				className="wc-block-products-grid wc-block-products-category"
			>
				{ __( 'No products were found that matched your selection.', 'woo-gutenberg-products-block' ) }
			</Placeholder>
		);
	}
```

## Screenshots

The above example let's me render this when my block renderer returns empty content:

![Edit Post ‹ local wordpress test — WordPress 2019-07-10 16-27-25](https://user-images.githubusercontent.com/90977/60982246-a8a9c700-a32f-11e9-8ee7-11305c63b5f2.png)

The default state is left alone if you don't pass a callback:

![Edit Post ‹ local wordpress test — WordPress 2019-07-10 16-29-16](https://user-images.githubusercontent.com/90977/60982373-e870ae80-a32f-11e9-98f8-aed8d31e77ea.png)

## Types of changes
This adds some non-required props to `<ServerSideRender>` so it shouldn't impact existing code.

## Checklist:
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->

mikejolley committed:
Add callbacks to ServerSideRenderer to handle failures

jorgefilipecosta commented on packages/server-side-render/src/server-side-render.js:
```diff
@@ -83,17 +83,17 @@ export class ServerSideRender extends Component {
+		const { className, onEmptyResponse, onErrorResponse, onNullResponse } = this.props;
 		if ( response === '' ) {
-			return (
+			return !! onEmptyResponse ? onEmptyResponse( this ) : (
```
Can onEmptyResponse be just a react element, instead of a function ? e.g: !! `onEmptyResponse ? onEmptyResponse : (`

Soean replied:
In the RSS Block we want to disable the block settings, if we get an error from the SSR. So I think a function is more flexible.

mikejolley replied:
I also thought function would be more flexible but open to change if it's required to get this through. Worth mentioning, the code using this component has no access to this components state without a function (passing `this`) which may be important, or not :)

jorgefilipecosta replied:
The case @Soean provided is interesting:
> In the RSS Block we want to disable the block settings.

We call the function onEmptyResponse on the render phase, render can not have side effects. So, for example, if there is a need to change the local state when the response is empty, that's not possible.
I guess to disable block settings on empty response, we would need to change a local state flag, so this change would not fit the RSS block use case.
I guess needing side effects when the response is empty will be common.
Maybe a possibility that would fit the use case would be an onEmptyResponse event. This function would be called, after the fetch is complete inside the fetch function, with an event object we define as a parameter. As a response to this event, users can apply any side effect desired e.g.: change state. This function would not render anything.
For the render, we have some possibilities:
	- have a prop emptyResponse that is rendered when the response is empty.
	- render nothing if onEmptyResponse is passed, the parent as a response to this event would be responsible for rendering the UI needed.

Any thoughts on this @mikejolley, @Soean?

mikejolley replied:
I've no thoughts on this one as my use case didn't require changing state, just acting on the state the block already had.

nerrad replied:
In looking at the comments here I like some of what @jorgefilipecosta proposes.  Here's what I'd like to see that I think will satisfy potential requirements for handling side-effects and a custom empty response render

-  Instead of an empty response event callback fired on completed fetch (as suggested by Jorge), I think it may be more extensible to simply fire any `fetchComplete` callback that is provided in props.  This opens up the possibility for components to trigger side-effects on various fetch states (empty response, successful response, caught fetch error).  This could be available for more advanced use-cases.
-  Just accept render props for the different response state props (and since these are placeholders, let's explicitly call them that) `EmptyResponsePlaceholder`, `NullResponsePlaceholder` and `ErrorResponsePlaceholder` options here and feed any state to them as they are rendererd.  You can then have default values for the render props (which could simplify some of the conditionals here).

jorgefilipecosta commented on packages/server-side-render/src/server-side-render.js:
```diff
@@ -83,17 +83,17 @@ export class ServerSideRender extends Component {
 				<Placeholder
 					className={ className }
 				>
 					{ __( 'Block rendered as empty.' ) }
```
I wonder if onEmptyResponse should be rendered inside the placeholder? The advantage is that it is easier to use, one does not need to pass the placeholder. The disadvantage is that it is not as customizable.

mikejolley replied:
I did it this way so we can pass a custom classname to Placeholder, or allow flexibility to use something completely different.

jorgefilipecosta commented on packages/server-side-render/src/server-side-render.js:
```diff
@@ -83,17 +83,17 @@ export class ServerSideRender extends Component {
 			);
 		} else if ( ! response ) {
-			return (
+			return !! onNullResponse ? onNullResponse( this ) : (
```
What is the relevant use case for a null response? Can onEmptyResponse handle all falsy values?

mikejolley replied:
Looking at the code, it seems the `! response` is ran when response is null. This seems to happen when things have not yet loaded? It shows a spinner. I was thinking by allowing each of these to have their own callback you could render a custom loading state too.

Does this need revising @jorgefilipecosta ?

nerrad replied:
Ya it looks like null is used to indicate a "loading" state. So there's value in leaving it as is.  However, I would suggest a slight name change here to make the public exposed api a bit clearer.  So instead of `NullResponse` for the render prop (continuing my earlier comment for the review), maybe make it `LoadingResponsePlaceholder`.

jorgefilipecosta reviewed (commented):
Thank you for your contribution @mikejolley 👍  Generally, it seems a good idea to add options to configure empty/error states.

mikejolley committed:
Update variable naming and set default renderers

mikejolley committed:
LoadingResponsePlaceholder

mikejolley committed:
swtich case

mikejolley committed:
fetchComplete

mikejolley commented:
@nerrad @jorgefilipecosta I've updated this with the suggested callbacks:

- emptyResponsePlaceholder
- errorResponsePlaceholder
- loadingResponsePlaceholder

The defaultProps handle the default placeholders which keeps the code in render() clean.

I also added `fetchComplete` as suggested but I'm not sure if thats how you meant to do it? I used `.finally` in the fetch request.

Let me know if this is looking better. I've also updated the PR in Woo which is using this (https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/717).

nerrad commented on packages/server-side-render/src/server-side-render.js:
```diff
@@ -123,4 +107,40 @@ export class ServerSideRender extends Component {
 	}
 }
 
+ServerSideRender.defaultProps = {
```
`defaultProps` is a feature of prop-types which we're not using in the Gutenberg project.  Instead, defaults should be set inline on the component.  You could probably define the defaults as their own functional components in this file and then just use them if the incoming render props are not set.

mikejolley replied:
@nerrad Can you check this again? I definitely found several files using `defaultProps`, and this code didn't break in testing?

nerrad replied:
oh huh!  I learned something new I guess :). I always thought it was a feature available via prop-types (basically because of it's mention in [this doc](https://reactjs.org/docs/typechecking-with-proptypes.html)) but it makes sense it's a native feature to React.  Move along then, nothing to see here 😜

nerrad commented on packages/server-side-render/src/server-side-render.js:
```diff
@@ -77,39 +77,23 @@ export class ServerSideRender extends Component {
-					{ __( 'Block rendered as empty.' ) }
-				</Placeholder>
-			);
+			return emptyResponsePlaceholder( this.props, response );
```
I don't think this should be a callback.  Just make it a render prop.  So consumers would provide the component they want rendered here and it would be implemented via:

```jsx
<emptyResponsePlaceholder response={ response } { ...this.props } />
```

or more conventionally via (which is more similar to what you have):

```js
onEmptyResponsePlaceholder( { ...this.props, response } );
```

mikejolley replied:
Take a look now - I just made this change. I renamed the props because the component names required a capital.

mikejolley commented on packages/server-side-render/src/server-side-render.js:
```diff
@@ -77,39 +77,23 @@ export class ServerSideRender extends Component {
 				}
+			} )
+			.finally( () => {
+				fetchComplete( this );
```
👍 Fine with me, I'll remove this from the PR.

mikejolley committed:
Remove fetchcomplete method

mikejolley committed:
Pass render props

nerrad reviewed (approved):
Looks good to me 👍 cc @jorgefilipecosta can you give this another looksee too?

nerrad committed:
add changelog entry

nerrad committed:
add documentation for new props
